cleverhans/attacks_tf.py

```diff
@@ -726,3 +726,135 @@ def compare(x, y):
+                   max_iter, clip_min, clip_max, nb_classes, feed=None):
+    """
+    Applies DeepFool to a batch of inputs
+    :param sess: TF session
+    :param x: The input placeholder
+    :param pred: The model's sorted symbolic output of logits, only the top
+                nb_candidate classes are contained
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients of the top nb_candidate classes, procuded
+                 from gradient_graph
+    :param X: Numpy array with sample inputs
+    :param nb_candidate: The number of classes to test against, i.e.,
+                        deepfool only consider nb_candidate classes when
+                        attacking (thus accelerate speed)
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for DeepFool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :param nb_classes: Number of model output classes
+    :return: Adversarial examples
+    """
+    X_adv = np.zeros(X.shape)
+    X_adv = deepfool_attack(sess, x, pred, logits, grads, X, nb_candidate,
+                            overshoot, max_iter, clip_min, clip_max, feed=feed)
+
+    return np.asarray(X_adv, dtype=np.float32)
+
+
+def deepfool_attack(sess, x, predictions, logits, grads, sample, nb_candidate,
+                    overshoot, max_iter, clip_min, clip_max, feed=None):
+    """
+    TensorFlow implementation of DeepFool.
+    Paper link: see https://arxiv.org/pdf/1511.04599.pdf
+    :param sess: TF session
+    :param x: The input placeholder
+    :param predictions: The model's sorted symbolic output of logits, only the
+                       top nb_candidate classes are contained
+    :param logits: The model's unnormalized output tensor (the input to
+                   the softmax layer)
+    :param grads: Symbolic gradients of the top nb_candidate classes, procuded
+                 from gradient_graph
+    :param sample: Numpy array with sample input
+    :param nb_candidate: The number of classes to test against, i.e.,
+                        deepfool only consider nb_candidate classes when
+                        attacking (thus accelerate speed)
+    :param overshoot: A termination criterion to prevent vanishing updates
+    :param max_iter: Maximum number of iteration for DeepFool
+    :param clip_min: Minimum value for components of the example returned
+    :param clip_max: Maximum value for components of the example returned
+    :return: an adversarial sample
+    """
+    import copy
+
+    adv_x = copy.copy(sample)
+    # Initialize the loop variables
+    iteration = 0
+    current = utils_tf.model_argmax(sess, x, logits, adv_x, feed=feed)
+    if current.shape == ():
```

Comment:
I read your response on the other PR, thanks. So my understanding is this requires that x is an image (i.e., has 4 dimensions). We'd like to have the attacks work on inputs of any shape as much as possible. Do you think the code could be rewritten to not depend on the shape of an image and take any input with a shape whose first dimension is the batch size?

Reply:
For this function, it does not require that x must have 4 dimensions (you can see it can work on random input with shape [100, 2] at test_attacks.py). The reason to judge the shape of current is due to the output of function utils_tf.model_argmax. When batch_size = 1, utils_tf.model_argmax will return a value with shape = (), while batch_size > 1, utils_tf.model_argmax will return a value with shape = (batch_size, ). In order to let the output under these two situations are in the same shape, we let current = np.array([current]) when current.shape == () (or we can replace this condition by batch_size == 1)